src/v/strings/static_str.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include <fmt/format.h>
+
+#include <string_view>
+#include <type_traits>
+
+#pragma once
+
+/**
+ * @brief A wrapper around a compile-time constant string.
+ *
+ * A static_str wraps an compile-time constant string, and is intended to be
+ * used at runtime as a guarantee that a string has static lifetime. Such string
+ * have trivial and cheap copy operations, since we just copy pointers (or
+ * views) to the underlying string.
+ */
+class static_str {
+public:
+    /**
+     * @brief Create a static_str based on a compile-time literal.
+     *
+     * The use of consteval prevents passing anything other than compile-time
+     * values.
+     */
+    consteval static_str(const char* str) // NOLINT
+      : _data{str} {
+        // do a compile-time check that the string is zero-terminated
+        std::string_view{str};
+    }
+
+    constexpr bool operator==(const static_str& rhs) const {
+        if (!std::is_constant_evaluated() && _data == rhs._data) {
+            // fast path based on pointer comparison, which is a sufficient but
+            // not necessary condition for equality
+            return true;
+        }
+        return *this == static_cast<std::string_view>(rhs);
+    }
+
+    constexpr auto operator<=>(const static_str& rhs) const {
+        if (!std::is_constant_evaluated() && _data == rhs._data) {
+            return std::strong_ordering::equal;
+        }
+        return *this <=> rhs.operator std::string_view();
+    }
+
+    /**
+     * @brief Return the static_str as a string_view.
+     */
+    constexpr operator std::string_view() const { return _data; } // NOLINT
+
+private:
+    // We could validly store the underlying string as a const char* or
+    // std::string_view, while offering the same API. We choose the pointer
+    // approach because it optimizes for size and the cost of copying, which is
+    // common on the fast path. It is slower to calculate the size, but the
```

Comment:
I did agonize about this decision. What should X be?

Some thoughts, though I don't necessary disagree (I started with `string_view`):

 - `string_view` is pretty much strictly worse for the use I'm introducing, which is passing this around a lot but never iterating the string even once (unless `trace` log is on)
 - see the update godbolt link for differences
 - string_view makes `size()` O(1) vs O(n) which could be a "big deal because factor n difference", I don't think that's a common case: usually you are consuming the entire string anyway, so it's more like the difference between being able to use `memcpy` vs `strcpy`, the latter also needing to check null termination "on the fly": then it's a QoI issue: sometimes both are a simple loop and similarly slow, sometimes only the known-size case is optimized to memcpy-alike but the null-terminated isn't and there's a big difference, sometimes both get optimized impls (e.g., strcpy which may read beyond the string on the source side but aligned so it never crosses a page), and then they can be close again (but the known size generally has a non-zero edge here).
 - If the compiler can see though to the source string, which will be common, it can of course calculate all this at compile time, e.g.,: https://godbolt.org/z/WjdosoMK6 (though this also means that the advantages of const char also melt away)
 - 2x larger is a non-trivial size increase if this is in a small structure (this does not occur in my use case)

So I'm torn.

A future use-case might end up stuffing these into hash tables or something or using them as lookup keys, and then the considerations are kind of reversed.

Reply:
Note there's an implementation variation which tries to get the best of both worlds: the `static_str` has a single pointer to a constexpr `string_view` object. This gets all the copying etc benefits of "single pointer" while having a known size. There's a small cost of an extra indirection when accessing the string.

However, this requires a different implementation AFAIK, since you need to generate this `constexpr static` string_view, which means the constructor needs to be a template, so you'd use something like `static_string<"abc">` to create one (and for sure you'd want a UDL to make this cleaner).

Reply:
> If this was `small_string_only_used_in_trace_logs` then I wouldn't even think about it. I guess what I am saying for a utility class in "string" I'd expect it to be optimized for wide usage rather than a very narrow one.

I don't think it's a slam dunk that general use is better off with `string_view` for this class, even though I agree that that is conventional wisdom.

> Maybe I am missing something but the only way to consume it is to convert it to string_view first via the conversion operator, right? So in that case you first do a strlen and then iterate again (aka two iterations instead of one).

OK yeah, so you caught me. This is true. I only added the minimal API that I thought was useful.

I guess you convinced me to base this on string_view. We can always change it later.

Reply:
Changed to `string_view` in https://github.com/redpanda-data/redpanda/commit/ed29c379e11385e9cd5c452f23b00ac8ee899121

Reply:
could you use the 64-48=16 free bits in the 64 bit pointer to store the length and fallback to strlen if needed?